packages/react-components/react-utilities/src/compose/slot.ts

```diff
@@ -0,0 +1,117 @@
+  value: Props | SlotShorthandValue | SlotComponent<Props> | undefined | null,
+  options?: { required?: boolean } & SlotOptions<Props>,
+): SlotComponent<Props> | undefined;
+export function slotFromShorthand<Props extends UnknownSlotProps>(
+  value: Props | SlotShorthandValue | SlotComponent<Props> | undefined | null,
+  options: { required?: boolean } & SlotOptions<Props> = {},
+): SlotComponent<Props> | undefined {
+  const { required = false, defaultProps, elementType } = options;
+
+  if (value === null || (value === undefined && !required)) {
+    return undefined;
+  }
+  const metadata: SlotComponentMetadata<Props> = isSlot<Props>(value) ? value[SLOT_COMPONENT_METADATA_SYMBOL] : {};
+
+  if (elementType) {
+    metadata.elementType = elementType;
+  }
+
+  // casting is required here as SlotComponent is an uncallable function, not an object
+  const propsWithMetadata = {
+    ...defaultProps,
+    [SLOT_COMPONENT_METADATA_SYMBOL]: metadata,
+  } as SlotComponent<Props>;
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  if (typeof value === 'string' || typeof value === 'number' || Array.isArray(value) || isValidElement<any>(value)) {
+    propsWithMetadata.children = value;
+  } else if (typeof value === 'object') {
+    Object.assign(propsWithMetadata, value);
+    if (typeof value.children === 'function') {
+      metadata.renderFunction = value.children as SlotRenderFunction<Props>;
+      propsWithMetadata.children = defaultProps?.children;
+    }
+  }
+
+  return propsWithMetadata;
+}
+
+/**
+ * Creates a slot for a v9 component - equivalent to `slot` method, but with a more specialized parameter,
+ * main difference is the presence of `props` and `ref`,
+ * which are a more specialized types than `value` on `slot` method.
+ *
+ * @example
+ * ```ts
+ * // useDialogBody
+ * export const useDialogBody_unstable = (props: DialogBodyProps, ref: React.Ref<HTMLDivElement>): DialogBodyState => ({
+ *   root: slotFromProps<DialogBodySlots>({ props, ref, elementType: props.as ?? 'div' }),
+ * });
+ * ```
+ */
+export function slotFromProps<Slots extends SlotPropsRecord, Primary extends keyof Slots = 'root'>(
+  props: PropsWithoutRef<ExtractSlotProps<Slots[Primary]>>,
+  params: SlotOptions<ExtractSlotProps<Slots[Primary]>> & {
+    ref?: 'ref' extends keyof ExtractSlotProps<Slots[Primary]> ? ExtractSlotProps<Slots[Primary]>['ref'] : never;
+  } = {},
+): SlotComponent<ExtractSlotProps<Slots[Primary]>> {
+  const { defaultProps, elementType, ref } = params;
+  const filteredProps: SlotComponent<ExtractSlotProps<Slots[Primary]>> = getNativeElementProps(elementType as string, {
+    ...defaultProps,
+    ref,
+    ...props,
+  });
+  filteredProps[SLOT_COMPONENT_METADATA_SYMBOL] = { elementType };
+  return filteredProps;
+}
+
+export function isSlot<Props extends {}>(element: unknown): element is SlotComponent<Props> {
+  return Boolean((element as {} | undefined)?.hasOwnProperty(SLOT_COMPONENT_METADATA_SYMBOL));
+}
```

Comment:
This isn't quite right, since `resolveShorthand` is now adding the metadata. You'd have to check if `elementType` is set as well.

That said, it seems like having both the render function and the elementType on a metadata object makes the implementation more complicated than it needs to be:
1. The metadata object can't require `elementType` because it might be missing after the call to `resolveShorthand`. But once a slot is resolved, we know that `elementType` must be set. It would be ideal for that to be reflected in the types.
2. The return type of `resolveShorthand` can't easily be different from the return type of `slotFromShorthand`, since they're both adding a metadata object.
3. It requires being more careful when modifying the metadata object, to be sure that you're not overwriting something that may have been added by an earlier call to `resolveShorthand`. Granted, that is all internal to the `slotFromShorthand` function and can be tested.

For all of those reasons (especially 1), it seems like it would be better to not have a metadata object, but instead just have two separate symbol props for the render function (existed before this PR) and element type (would be new in this PR). In that case, the `isSlot` function would only check for the existence of the element type prop.

What's the reason for combining them both on a metadata object?